Template integration tests

matthew-rollitt opened the pull request:
This branch adds the automated testing to the templates
important files:
Runner/
'Runner.py'
'JobManager.py' responsible for creating, submitting and monitoring the pools and jobs
'CustomTemplateFactory.py' is for reading and setting elements of the json templates
'Utils.py' module that holds the data objects and some useful methods
'ReadMe.md'

There are some rendering files being added to this repo that are in the /Assets folder 

to run the program please use python .\Runner.py ".\Tests\TestConfiguration.json"

matthew-rollitt committed:
Can create pools/jobs

First commit before going to await/async pattern. Currently creates all the pools and jobs and checks to see if the output file is correct

matthew-rollitt committed:
Started breaking out jobs into their own class

matthew-rollitt committed:
Merge branch 'support-latest-image' into batchexplorer-runner

matthew-rollitt committed:
Merge branch 'support-latest-image' into batchexplorer-runner

Added more job types and started to remove all the old code

matthew-rollitt committed:
Vray standalone and Arnold standalone now works

matthew-rollitt committed:
ImageMagick job

matthew-rollitt committed:
Big refactor

Got some advice and changed the way we handled creating the job

matthew-rollitt committed:
Merge branch 'support-latest-image' into batchexplorer-runner

matthew-rollitt committed:
Added vray tests

matthew-rollitt committed:
More tests added

Added

matthew-rollitt committed:
Merge branch 'support-latest-image' into batchexplorer-runner

matthew-rollitt committed:
Merge branch 'master' into batchexplorer-runner

smith1511 committed:
VRay DR Rendering fixes

smith1511 committed:
Update URLs to master branch

matthew-rollitt committed:
Updated runner

Now works with a config file. Fixing some async stuff,

matthew-rollitt committed:
Added Template Runner

Small program that can consume a json file and create and tests jobs/pools on them. If the job fails the container should be left over

matthew-rollitt committed:
Updated async and a few other things

Fixed the async behavior that was causing null errors. Fixed json files so containers can be created correctly. Removed a bunch of Libs that won't be needed and cleaned up a few other methods based on a PR

matthew-rollitt committed:
More tests passing, better error handling for failed cases

matthew-rollitt committed:
Merge branch 'master' into template-runner

matthew-rollitt committed:
handling retry logic nicely. Added the ability to select the image you want in the config file. Cleaned up output of the runner.

matthew-rollitt committed:
Merge branch 'master' into batchexplorer-runner

matthew-rollitt committed:
Merge branch 'master' into template-runner

matthew-rollitt committed:
Updating yaml file

matthew-rollitt committed:
testing

Testing yaml file

matthew-rollitt committed:
Missed runner py

matthew-rollitt committed:
Updating requirements link

matthew-rollitt committed:
Update azure-pipelines.yml

testing env vars

matthew-rollitt committed:
Adding env vars

matthew-rollitt committed:
Added import os

matthew-rollitt committed:
Missing PS_SERVICE_PRINCIPAL_CREDENTIALS_SECRET

matthew-rollitt committed:
fixing path to testconfig

matthew-rollitt committed:
Handle ubuntu and windows paths

matthew-rollitt committed:
Debugging issue with filepath

matthew-rollitt committed:
Removed debugging

matthew-rollitt committed:
Small doc change

matthew-rollitt committed:
testing env vars

matthew-rollitt committed:
testing Junit

matthew-rollitt committed:
Testing junit #1

matthew-rollitt committed:
Removed user input

matthew-rollitt committed:
Trying to get the outputfile working

matthew-rollitt committed:
printing current folder

matthew-rollitt committed:
Fixed path in the yaml file

matthew-rollitt committed:
Runner not creates a juit export file

matthew-rollitt committed:
removed unneeded code

matthew-rollitt committed:
cleaning up output

matthew-rollitt committed:
testupdate

matthew-rollitt committed:
not handling error correctly

matthew-rollitt committed:
adding a try/catch

matthew-rollitt committed:
updated python version

matthew-rollitt committed:
Only testing non preview images

matthew-rollitt committed:
More testing

matthew-rollitt committed:
removed _time

matthew-rollitt committed:
Time should be working now

matthew-rollitt committed:
handling job-id vs time-job-id

matthew-rollitt committed:
testing config 1

matthew-rollitt committed:
Only keeps the containers if the pool resized and correctly

matthew-rollitt committed:
Added more timeout  handlers

matthew-rollitt committed:
Fixed reading timeout issue

matthew-rollitt committed:
udpated timeout to 30min

matthew-rollitt committed:
Merge branch 'vraydr-fixes-final' into template-runner-yaml-build

matthew-rollitt committed:
Changed timeout

matthew-rollitt committed:
increased the version of windows

matthew-rollitt committed:
Fixed the stopwatch

matthew-rollitt committed:
Updating timeout logic

matthew-rollitt committed:
updated time

matthew-rollitt committed:
updatetimeout

matthew-rollitt committed:
updated timeout to be more precise

matthew-rollitt committed:
Fixed total time, needed to return the total seconds.

matthew-rollitt committed:
Merge branch 'new-windows-images' into template-runner-yaml-build

matthew-rollitt committed:
fixed job.params file for blender linux job

matthew-rollitt committed:
Moved export into finally

matthew-rollitt committed:
Added clean up  old containers

matthew-rollitt committed:
Remove raise

matthew-rollitt committed:
delete pool at the end

matthew-rollitt committed:
Small clean up.

matthew-rollitt committed:
Merge branch 'template-runner-yaml-build' of https://github.com/Azure/BatchExplorer-data into template-runner-yaml-build

matthew-rollitt committed:
Merge branch 'template-runner-yaml-build' into batchexplorer-runner

matthew-rollitt committed:
Small test change

matthew-rollitt committed:
Merge branch 'master' into batchexplorer-runner

matthew-rollitt committed:
fix upload

matthew-rollitt committed:
Run prep8 and did some refactoring.

matthew-rollitt committed:
Merge branch 'master' into batchexplorer-runner

matthew-rollitt committed:
Added some more doc strings

matthew-rollitt committed:
Added clean up code back in

matthew-rollitt committed:
removed a few unneeded values in json

matthew-rollitt committed:
Increased timeout

matthew-rollitt committed:
Updated read me

matthew-rollitt committed:
Delete test file

timotheeguerin commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+
+    # Set output filegroup
+    try:
+        template["outputFilegroup"]["value"] = storage_info.output_container.replace("fgrp-", "")
```
could the container have `frgp-` somewhere else in the name. If so maybe using a regex might be better

matthew-rollitt replied:
frgp- needs to be the prefix.

timotheeguerin commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+
+    # If template is windows version
+    if "windows" in template_image_reference["offer"]:
+        for i in range(0, len(image_references)):
```
why not using `for image_ref in image_references` instead

timotheeguerin replied:
same below

matthew-rollitt replied:
I'll make this small change

timotheeguerin commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+    job_id = ""
+    with open(parameters_file) as f:
+        parameters = json.load(f)
+    try:
```
shouldn;'t that be inside the with?

timotheeguerin replied:
same for the other functions below

matthew-rollitt replied:
I'll make this change.

timotheeguerin commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+            job_json)
+        job = await loop.run_in_executor(None, functools.partial(batch_service_client.job.add, jobparameters))
+    except batchmodels.batch_error.BatchErrorException as err:
+        print(
```
try to use `logging` instead of print

matthew-rollitt replied:
Yep switching every print statement over to use logging

timotheeguerin commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+        all_pools = [p.id for p in await loop.run_in_executor(None, functools.partial(batch_service_client.pool.list))]
+
+        ctm.set_template_name(template, self.pool_id)
+        if(self.pool_id not in all_pools):
```
no need for (

matthew-rollitt replied:
removed

timotheeguerin commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+        output_container_name = "fgrp-" + self.job_id + '-output'
+
+        # Create input container
+        await loop.run_in_executor(None, functools.partial(blob_client.create_container, input_container_name, fail_on_exist=False))
```
why do you need to do that? await + run_in_executor

timotheeguerin commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+            self.job_status = Utils.JobStatus(Utils.JobState.POOL_FAILED,
+                                              "Job failed to start since the pool [{}] failed to allocate any TVMs due to error [Code: {}, message {}]."
+                                              .format(self.pool_id, pool.resize_errors[0].code, pool.resize_errors[0].message))
+            print("POOL {} FAILED TO ALLOCATE".format(self.pool_id))
```
logging

matthew-rollitt replied:
Yep, will do

timotheeguerin commented on Runner/ReadMe.md:
```diff
@@ -0,0 +1,45 @@
+This python script is used for testing the Batch Templates inside the BatchExplorer-data\ncj folder. It does not test BatchExplorer only the Azure Batch CLI and its extension client. Please set the following environmental variables before running the job script. 
+
+PS_BATCH_ACCOUNT_NAME
+PS_BATCH_ACCOUNT_KEY
```
need this key? and the storage account key bellow? should be able to get it from arm

timotheeguerin commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+    print("_BATCH_ACCOUNT_URL", _BATCH_ACCOUNT_URL)
+    print("_BATCH_ACCOUNT_SUB", _BATCH_ACCOUNT_SUB)
+    print("_STORAGE_ACCOUNT_NAME", _STORAGE_ACCOUNT_NAME)
+    print("_STORAGE_ACCOUNT_KEY", _STORAGE_ACCOUNT_KEY)
```
shouldn't print the credentials

timotheeguerin replied:
you should reset those keys too now

ascobie replied:
Have talked to Matt about this and he is going to put the credentials in KeyVault as storing account keys in environment variables is a security risk.

timotheeguerin replied:
that's a good idea I think VSTS has good integration with keyvault

matthew-rollitt replied:
Yep that's the plan, I also need to sort out what account I can use.

matthew-rollitt replied:
I'll add logging that will print out the account name and the storage account being used, that's all

matthew-rollitt replied:
Yep I'm using keyvault  now

timotheeguerin commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+
+        print("{} jobs will be created".format(len(_job_managers)))
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(asyncio.gather(*[j.upload_assets(blob_client) for j in _job_managers]))
```
why all those loop why not use await?

timotheeguerin commented on azure-pipelines.yml:
```diff
@@ -7,10 +7,56 @@ pool:
-  displayName: 'Run a one-line script'
+- task: UsePythonVersion@0
+  inputs:
+    versionSpec: 3.6
```
upgrade to 3.7

matthew-rollitt replied:
will do

matthew-rollitt replied:
Side note 3.7.1 doesn't work. It's unsupported at this time

timotheeguerin replied:
unsupported by azure devops?

timotheeguerin commented on azure-pipelines.yml:
```diff
@@ -7,10 +7,56 @@ pool:
+ 
+      Runner_py = next(iglob(path.join('**', 'Runner.py'), recursive=True), None)
+      if not Runner_py:
+          raise SystemExit('Could not find a Django project')
```
? what is this doing?

matthew-rollitt replied:
basically saying if we can't find the runner.py file throw an error.

Tatsinnit replied:
💡 I think what Tim is coming at is the error message says a word called : `django` Django is a different py `framework` read here: ``https://www.djangoproject.com/ 

* The error raised is utilized for some specific purposes, I woul recommend add more info like: `Could not found project {}.format(proj_name)` or something.

timotheeguerin replied:
wondering why it couldn't find this file here though?

Tatsinnit replied:
I have a feeling that **use** of `async` and `exception handling` is vastly used in these files as loose concepts? (unsure why) @matthew-rollitt Please add comments and Details at places where things are not clear. 

2 things will benefit:

* We will be able to pin point or recommend if you are not sure.
* You will be clear how the runner is working end-to-end.

timotheeguerin commented on azure-pipelines.yml:
```diff
@@ -7,10 +7,56 @@ pool:
+      import os
+      # Python >= 3.6
+
+      os.environ["PS_BATCH_ACCOUNT_KEY"] = "$(PS_BATCH_ACCOUNT_KEY)" 
```
why do you need that for all?

matthew-rollitt replied:
I'll remove them

matthew-rollitt replied:
so this should be changed to keyvault. but basically i was loading all the configs into the env and reading them that way.

timotheeguerin reviewed (changes requested):
- [ ] Can you change the files names to follow python pythono convention all lowercase with underscore
- [ ] Can you enable pylint to check for syntax errors 

This is the config we use for aztk, we removed rules which are just anoying https://github.com/Azure/aztk/blob/master/pylintrc

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+    print("_STORAGE_ACCOUNT_KEY", _STORAGE_ACCOUNT_KEY)
+    print("_SERVICE_PRINCIPAL_CREDENTIALS_CLIENT_ID",
+          _SERVICE_PRINCIPAL_CREDENTIALS_CLIENT_ID)
+    print("_SERVICE_PRINCIPAL_CREDENTIALS_SECRET",
```
📓 :bulb: What is the purpose of these `Prints`?  are these only debug purpose prints if so mention it and please remove before final checkout.

IMO: I think no need to add these prints in final version et. al.

matthew-rollitt replied:
I'll add logging that will print out the account name and the storage account being used, that's all

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+        with open(TestConfigurationFile) as f:
+            template = json.load(f)
+
+        for i in range(0, len(template["tests"])):
```
:bulb: 🔔  This seems like `for each` in need here, rather then this:

Please see this:

```
for val in array:
    print(val)

``` 
So in your code you can do:

```
for val in template["tests"]):
    var_val = val
```

matthew-rollitt replied:
I'll make this change.

matthew-rollitt replied:
ah thanks

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+            template = json.load(f)
+
+        for i in range(0, len(template["tests"])):
+            test = template["tests"][i]
```
:bulb: rename variables to more `logical name` sets please. i.e. `test` is hyper common.

matthew-rollitt replied:
how about 'jobSettings'?

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+
+            applicationLicenses = None
+            try:
+                applicationLicenses = test["applicationLicense"]
```
💡 This `try` will throw what? i.e. the purpose of `try-except` is to make sure we capture exception. Now looking at this specific `try` it only contains `assignment`, what is the reason for this try for assignment? (just wondering)

matthew-rollitt replied:
In the test config file 'applicationLicenses' is optional inside the Json. If it's not set the value will be a none and it will throw a KeyError. If that happens we just catch the error and move on. If we get another error we should throw it.  

I don't see a nicer way of handling this

timotheeguerin replied:
I think you can also do that https://stackoverflow.com/questions/25833613/python-safe-method-to-get-value-of-nested-dictionary but I'm not sure its better

matthew-rollitt replied:
It doesn't look any better.

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+            except KeyError as e:
+                pass
+            except Exception as e:
+                raise e
```
💡 please refer to above comments and the actual cause of assignment exception is unexplained, so not sure if this is in need.

matthew-rollitt replied:
yep removed all the keyError

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+                    applicationLicenses))
+
+        images_refernces = []
+        for i in range(0, 0):
```
🔔 💡 Hardcoded range? (this is incorrect)

matthew-rollitt replied:
oops ... yeah that needs to be for i in template[images]

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+_job_managers = []
+
+
+if __name__ == '__main__':
```
🔔 This whole is in one massive init funct I think can we please modularise it a bit i.e. `cleate a funct` and call at `init`?

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+
+
+if __name__ == '__main__':
+    start_time = datetime.datetime.now().replace(microsecond=0)
```
🔔 this might help you for the purpose what you are trying to do here: https://stackoverflow.com/questions/7999935/python-datetime-to-string-without-microsecond-component

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+        raise e
+
+    try:
+        with open(TestConfigurationFile) as f:
```
🔔 Word `test` naming convention is very loosely used please reconsider like : `jsonConfFile` or something.

matthew-rollitt replied:
i'll make a change

Tatsinnit commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+    try:
+        template["jobId"]["value"] = job_id
+    except KeyError:
+        pass
```
🔔 same statement repeated `try-except` many times, i.e. I think copy paste error? Why not both `statement` under 1 try block? Also, Redundant `exception` handle.

timotheeguerin replied:
Think this is to make sure if the first one fail the second actually still runs. Also those try except are needed they if ignore the exception which is what is wanted here

matthew-rollitt replied:
I've changed this up

Tatsinnit commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+
+
+def load_file(template_file: str) -> str:
+    template = ""
```
💡 I think this is not needed.

matthew-rollitt replied:
Nope load_file is used in create_and_submit_Job

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+            template, parameters)
+        jobparameters = batch_service_client.job.jobparameter_from_json(
+            job_json)
+        job = await loop.run_in_executor(None, functools.partial(batch_service_client.job.add, jobparameters))
```
🔔 Curious: Why `executor` thread here? but not `run until complete` **or** simple `await`?

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+                filePath = Path("Assets/" + file)
+                await loop.run_in_executor(None, functools.partial(Utils.upload_file_to_container, blob_client, input_container_name, filePath))
+
+    async def check_expected_output(self, batch_service_client):
```
FYI: mostly you also add `Param and return type` in comments.

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+            return False
+
+        # Wait for TVMs to become available
+        nodes = list(batch_service_client.compute_node.list(self.pool_id))
```
🔔 seem like this : `batch_service_client.compute_node.list(self.pool_id)` is already a **List** so why we double cast it? like `list(list)`?

matthew-rollitt replied:
nice, don't the list()

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+        print("Submitting jobs...")
+        loop.run_until_complete(asyncio.gather(*[j.create_and_submit_Job(batch_client) for j in _job_managers]))
+        print("waiting for jobs to complete...")
+        loop.run_until_complete(asyncio.gather(*[j.wait_for_tasks_to_complete(batch_client, _timeout) for j in _job_managers]))
```
Use of `async` logic should be revisited: `await` can be used but the `run until complete` should work, but what is the need of a blocking method which essentially has no `wait` ==> `wait_for_tasks_to_complete` is `non-async` but name suggest it is.

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+        # Reties any jobs that failed
+        print("-----------------------------------------")
+        loop.run_until_complete(asyncio.gather(*[j.retry(batch_client, blob_client, _timeout/2) for j in _job_managers]))
+        loop.run_until_complete(asyncio.gather(*[j.delete_resouces(batch_client, blob_client) for j in _job_managers]))
```
🔔 Why do we need `asyncio.gather`?

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,155 @@
+
+        end_time = datetime.datetime.now().replace(microsecond=0)
+        Utils.export_result(_job_managers, (end_time - start_time))
+    print()
```
minor : empty print?

matthew-rollitt replied:
I was using this for formatting.

Tatsinnit replied:
💡 Cool, Please refer to other places for consitent formatting in this case: like in here: https://github.com/Azure/BatchExplorer-data/pull/28/files#diff-a4776f1fa82c7a5bd1d4df3f9fdb662bR143

matthew-rollitt replied:
yep removed all empty prints after switching to logging

Tatsinnit commented on Runner/Tests/TestConfiguration2.json:
```diff
@@ -0,0 +1,45 @@
+{
+  "tests": [
```
⏺ Possible contextual rename required here.

matthew-rollitt replied:
not going to make this change. had a chat with Ivan over this.

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,284 @@
+    POOL_FAILED = 2
+    # Job ran to completion and the output matched the test configuration file
+    COMPLETE = 3
+    # The outout file did not match the expected output desscibed in the test
```
**Note** Minor: output not `outout` I think: this file is better commented then others I would recommend doing similar style of commenting in other py file

matthew-rollitt replied:
Good catch, If you see something that is lacking in documentation  that needs to be highlighted let me know

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,334 @@
+        # wait for pool to come up
+        while pool.allocation_state.value == "resizing" and self.check_timeout(
+                timeout):
+            time.sleep(10)
```
:bulb: What is the purpose of `sleep` here? (Please add comment)

matthew-rollitt replied:
Queries the pool status every 10 seconds instead of every second. I don't want to query the service every second because it would be too noisy.

Tatsinnit commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+import json
```
📓 Please note if this is needed in your project: `# -*- coding: utf-8 -*-`

* https://docs.python.org/3/reference/lexical_analysis.html#encoding-declarations

matthew-rollitt replied:
will do

Tatsinnit reviewed (changes requested):
Initial review recommendation and questions.

Tatsinnit commented:
- [ ] 💡 **Recommendation for security scan** Once you think this all is ready, you should check your python files for `security` checks using this `bandit`: https://pypi.org/project/bandit/ (This will be local only check) but should give you a good scan.

matthew-rollitt commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,170 @@
+    try:
+        template["parameters"]["poolName"]["defaultValue"] = pool_id
+    except KeyError:
+        pass
```
I don't feel like thses are Redundant 

def set_template_name(template, pool_id):
    try:
        template["parameters"]["poolName"]["defaultValue"] = pool_id
    except KeyError:
        pass
    try:
        template["parameters"]["poolId"]["defaultValue"] = pool_id
    except KeyError:
        pass
    try:
        template["parameters"]["poolId"]["value"] = pool_id
    except KeyError:
        pass

vs this are very different 
def set_template_name(template, pool_id):
    try:
        template["parameters"]["poolId"]["value"] = pool_id
        template["parameters"]["poolName"]["defaultValue"] = pool_id
        template["parameters"]["poolId"]["defaultValue"] = pool_id
    except KeyError:
        pass

The point of these methods is to set a single value in the template, if that value isn't in there in the template then that means we failover and try and set a different value. From my understanding, the second example would fail if the first set fails and it won't proceed to the second set.

matthew-rollitt committed:
Updated code based on a PR

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,332 @@
+            template, parameters)
+        jobparameters = batch_service_client.job.jobparameter_from_json(
+            job_json)
+        job = await loop.run_in_executor(None, functools.partial(batch_service_client.job.add, jobparameters))
```
💡 `batch_service_client.job.add` is this a async function?

matthew-rollitt replied:
I've gone ahead and removed all async functions and replaced them with threading

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,332 @@
+            template, parameters)
+        jobparameters = batch_service_client.job.jobparameter_from_json(
+            job_json)
+        job = await loop.run_in_executor(None, functools.partial(batch_service_client.job.add, jobparameters))
```
Please note: Which ever prefered way you intend to choose for `async` keep it consistent. Since we are upgrading this to `Python 3.7` key update we can use now is the prefered way of handling `coroutine and tasks` https://docs.python.org/3/library/asyncio-task.html 

```
import asyncio
import time

async def say_after(delay, what):
    await asyncio.sleep(delay)
    print(what)

async def main():
    print(f"started at {time.strftime('%X')}")

    await say_after(1, 'hello')
    await say_after(2, 'world')

    print(f"finished at {time.strftime('%X')}")

asyncio.run(main())

```
**Expected output** :

```
started at 12:13:52
hello
world
finished at 12:13:55
```
For gather sample: (To run task concurrently) https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently 

```
import asyncio

async def factorial(name, number):
    f = 1
    for i in range(2, number + 1):
        print(f"Task {name}: Compute factorial({i})...")
        await asyncio.sleep(1)
        f *= i
    print(f"Task {name}: factorial({number}) = {f}")

async def main():
    # Schedule three calls *concurrently*:
    await asyncio.gather(
        factorial("A", 2),
        factorial("B", 3),
        factorial("C", 4),
    )

asyncio.run(main())

# Expected output:
#
#     Task A: Compute factorial(2)...
#     Task B: Compute factorial(2)...
#     Task C: Compute factorial(2)...
#     Task A: factorial(2) = 2
#     Task B: Compute factorial(3)...
#     Task C: Compute factorial(3)...
#     Task B: factorial(3) = 6
#     Task C: Compute factorial(4)...
#     Task C: factorial(4) = 24
```

matthew-rollitt replied:
I've gone ahead and removed all async functions and replaced them with threading. Found threading to be exactly what I needed to use and easier. Yes under the hood they are similar but the removing all async/await operations made my code easier to understand if someone was to read over the code. 

The whole idea is for every template we want to test we want to create a new Thread where it can be run and wait for it to complete. Sure we can do the same thing with async gather but it more straightforward using a thread pool

matthew-rollitt replied:
I'll be adding in some mock tests for the next chunk of work I need to do once this is in the new repo. I'll also need a offline template tester that will compare params json file vs the job json file since this will cause a lot of build issues later down the track. There are some plans to move this branch somewhere else I would like this branch in before the move.

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,332 @@
+        # Submits the job
+        await submit_job(batch_service_client, template, parameters)
+
+    async def create_pool(self, batch_service_client, image_references):
```
💡  Recommnedation: You can always add async to the name of the async funciton like here this funct can be : `create_pool_async`.

matthew-rollitt replied:
Yep i'll do this

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,332 @@
+            pool = batch_service_client.pool.poolparameter_from_json(pool_json)
+            Utils.logger.info('Creating pool [{}]...'.format(self.pool_id))
+            try:
+                await loop.run_in_executor(None, functools.partial(batch_service_client.pool.add, pool))
```
💡 I think it will be worth it to comment here, but pelase clarify is this funciton is `async` at all?

matthew-rollitt replied:
Will do

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,332 @@
+    str(datetime.datetime.now().minute)
+
+
+async def submit_job(batch_service_client, template, parameters):
```
📓 Please think about adding simple`Unit Test` for key functionality.

matthew-rollitt replied:
Haven't done this yet. But i'll do after we move over to the new repo.

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,141 @@
+        # Reties any jobs that failed
+        Utils.logger.info("-----------------------------------------")
+        loop.run_until_complete(asyncio.gather(*[j.retry(batch_client, blob_client, _timeout/2) for j in _job_managers]))
+        #loop.run_until_complete(asyncio.gather(*[j.delete_resouces(batch_client, blob_client) for j in _job_managers]))
```
💡 Please note these are commented lines, if no longer in need, then please delete them.

matthew-rollitt replied:
They need to be uncommented in the final PR, This is done to speed up testing.

Tatsinnit commented on Runner/Tests/TestConfiguration3.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "tests": [
```
📓 Possible contextual rename required here as well.

matthew-rollitt replied:
I talked with Ivan on this one, he agrees with 'tests'. Won't change

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,300 @@
+fh = logging.FileHandler('template.log')
+fh.setLevel(logging.DEBUG)
+# create console handler with a higher log level
+ch = logging.StreamHandler()
```
:bulb: **Recommendation** rename `ch` to `console_handler` or something more obvious.

matthew-rollitt replied:
okay, i'll also do fh as well

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,300 @@
+    logger.info("Exporting test output file")
+    root = Element('testsuite')
+
+    for i in job_managers:
```
**recommendation** rename `i` to more logical contextual name. might be : `job_item`.

matthew-rollitt replied:
Yep, i'll change it into job_item

matthew-rollitt committed:
Remove all async behavior and changed to a thread pool, made lots of changes based on the PR. Added a logger class to handle writing to log file more easier.

matthew-rollitt committed:
Adding loggign script

matthew-rollitt committed:
changing print

matthew-rollitt committed:
checking vars

matthew-rollitt committed:
test

matthew-rollitt committed:
change print

matthew-rollitt committed:
testing

matthew-rollitt committed:
remove

matthew-rollitt committed:
udpate

matthew-rollitt committed:
update

matthew-rollitt committed:
single line

matthew-rollitt committed:
small update

matthew-rollitt committed:
missed quote

matthew-rollitt committed:
pushd '$(projectRoot)'

matthew-rollitt committed:
reseting file

matthew-rollitt committed:
extra value

matthew-rollitt committed:
added delete pool

matthew-rollitt committed:
Merge branch 'pipeline-fix' into batchexplorer-runner

matthew-rollitt committed:
Merge branch 'master' into batchexplorer-runner

matthew-rollitt committed:
Updated a few params files that failed after the merge from master

Tatsinnit commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,187 @@
+# -*- coding: utf-8 -*-
+import json
```
💡 Please note module names should have all shorts: PEP 8 standards here: https://www.python.org/dev/peps/pep-0008/#package-and-module-names  ==> hence in this case right name convention for the module will be : `customtemplatefactory.py`

Tatsinnit commented on Runner/Logger.py:
```diff
@@ -0,0 +1,118 @@
+from xml.etree.ElementTree import Element, SubElement, ElementTree
+import logging
```
💡 same as above recheck file names for your python files and follow the standards mentioned here: https://www.python.org/dev/peps/pep-0008/#package-and-module-names

Tatsinnit commented on Runner/Logger.py:
```diff
@@ -0,0 +1,118 @@
+logger = logging.getLogger('rendering-log')
+logger.setLevel(logging.DEBUG)
+# create file handler which logs even debug messages
+fh = logging.FileHandler('template.log')
```
💡 we discussed to rename these to more logical and you mentioned that you have made all these renaming changes perhaps a missing commit? like `fh` = file_handler or something. just checking is a commit was dropped. At this point this is too minor for me.

matthew-rollitt replied:
Hmm thing my change got reverted out. I've put the change back in.

Tatsinnit commented on Runner/Runner.py:
```diff
@@ -0,0 +1,152 @@
+    """
+    Create a batch client using AAD.
+
+    :type args: ArgumentParser
```
💡 `return:` type param?

matthew-rollitt replied:
Thanks, I'll go through and check for other times I'm missing the return type

Tatsinnit commented on Runner/Utils.py:
```diff
@@ -0,0 +1,266 @@
+        # Grab all the tasks in the Job.
+        tasks = batch_service_client.task.list(job_id)
+        # tasks = yield loop.run_in_executor(None,
+        # functools.partial(batch_service_client.task.list,
```
💡 Please remove all this commented code if this is not needed?

matthew-rollitt replied:
Thanks, will do

Tatsinnit commented on Runner/CustomTemplateFactory.py:
```diff
@@ -0,0 +1,187 @@
+    """
+
+    # 'fgrp-' needs to be removed.
+    if 'inputData' in template:
```
💡 The `type` of `template` mentioned above is jsonfile? is so will the string comparison with `in` clause will work? or is it the wrong type in the `param` clause?

I think the `param is mentioned of wrong type. so in your case is a `jsonData` which can be checked using the `in` clause.

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,380 @@
+
+        # Upload the asset file that will be rendered and
+        scenefile = ctm.get_scene_file(self.parameters_file)
+        for file in os.listdir("Assets"):
```
📓 the path provided to the `listdir` is the base path of the file, i.e. in your case, think to remember is that the `Assets` is always going to be a set path, just mentioning for double checking.

matthew-rollitt replied:
Thanks for checking, for now, this is how I'm handling the files. This may change in the future

Tatsinnit commented on Runner/JobManager.py:
```diff
@@ -0,0 +1,380 @@
+                traceback.print_exc()
+                Utils.print_batch_exception(batch_exception)
+
+    def delete_resouces(self, batch_service_client, blob_client, force_delete=False):
```
🐛 name mistake I think you meant : `delete_resources` not `resouces`

matthew-rollitt replied:
good spot!

Tatsinnit replied:
All good! @matthew-rollitt I will drop a check list before you I am done with this, please test your scenarios, also look at the PEP 8 recommendations above.

timotheeguerin commented on Runner/ReadMe.md:
```diff
@@ -0,0 +1,45 @@
+
```
rename file to `README.md`

timotheeguerin commented on Runner/Logger.py:
```diff
@@ -0,0 +1,118 @@
+    Logs the account info
+    :type args: ArgumentParser
+    """
+    info("-----------------------------------------------------")
```
inconsitent as you use `logger.info` below

matthew-rollitt committed:
A lot of doc changes, annotations and files renamed

matthew-rollitt committed:
Renamed Runner.py to runner.py

matthew-rollitt committed:
Reformated file and updated a doc string

timotheeguerin commented on Runner/utils.py:
```diff
@@ -236,31 +251,32 @@ def cleanup_old_resources(blob_client):
 
 
-def create_thread_collection(method_name, job_managers, *args):
+def create_thread_collection(method_name: str, job_managers: 'list[job_manager.JobManager]', *args):
```
does this needs to be in quotes? 'list[job_manager.JobManager]'

matthew-rollitt replied:
https://www.python.org/dev/peps/pep-0484/#forward-references  used to prevent circular referencing

timotheeguerin commented on Runner/utils.py:
```diff
@@ -49,55 +51,62 @@ class JobStatus(object):
+    :type message: str
+    :return bool: if the expected exception is hit return a True.
     """
     if batch_exception.error and \
```
minor style. I think it looks nicer when you align the operators at the beginning
```py
if foo
  and bar
  and something:

```

timotheeguerin commented on Runner/logger.py:
```diff
@@ -1,68 +1,88 @@
 from xml.etree.ElementTree import Element, SubElement, ElementTree
```
also rename the `Runner` folder to `runner`

matthew-rollitt committed:
More doc changes and some renames

matthew-rollitt committed:
Removed workspace.xml file

Tatsinnit commented:
Note: glad that async is all removed and 4 things at this point will make this PR better: 

- [x] Docstrings, spell checks, rtype and object name well defined. 
- [x] Better explanation of the use of utils function: `create_thread_collection`
- [x] Consistent use of module naming and definition structure.
- [x] Fix current conflicts.

If you think all the 4 points are catered, feel free to get few folks to take a quick look.

**Important**

- [x] Whenever you get the greenlight : Please **Squash and Merge**.

Thanks,

matthew-rollitt committed:
Small changes

matthew-rollitt committed:
Renamed the Runner folder to runner

matthew-rollitt committed:
Merge branch 'master' into batchexplorer-runner

Tatsinnit commented on runner/utils.py:
```diff
@@ -0,0 +1,286 @@
+        raise e
+
+
+def create_thread_collection(method_name: str, job_managers: 'list[job_manager.JobManager]', *args):
```
💡 please add better explanation of this `Thread` logic here? I am coming from this opinion: https://stackoverflow.com/questions/43747261/when-to-use-threading-and-how-many-threads-to-use 

* Hence in current implementation I might have missed some thing, so add a scenario so that it will help to understand why `threads` are used here. 
* A scenario will help in understanding as well as a note for others. (Or if you retrospect)

matthew-rollitt committed:
logger fix, doc string update and CI fix

logger was logging warnings and errors to the console and the log file. Now it just logs to the log file. Fixed the merge issue with the templates and udpated some doc strings

matthew-rollitt committed:
fixed path

matthew-rollitt committed:
Fixed error message

matthew-rollitt committed:
missing s for some utils in docs

matthew-rollitt committed:
Added a pool timestamp. This prevents the issue of having multiple runners coliding one another.

matthew-rollitt commented:
This PR adds the ability to automatically test the job/pool templates inside the ncj/ folder. Whenever we have a new rendering image or update the rendering templates these tests will run on every build. If you want to add a new rendering scenario you just a test to the  runner/TestConfiguration.json file. 

This runner will run on every build and takes roughly 15-25min to run

Tatsinnit commented:
> This PR adds the ability to automatically test the job/pool templates inside the ncj/ folder. Whenever we have a new rendering image or update the rendering templates these tests will run on every build. If you want to add a new rendering scenario you just a test to the runner/TestConfiguration.json file.
> 
> This runner will run on every build and takes roughly 15-25min to run

Cool, Please note: this can very well be the `Squash commit message` with little bit more summarized. I think if you are happy with this, you can get guys to merge it with a quick eyeball. 

For history, here is the old-old PR as reference and how this work has grown etc: https://github.com/Azure/BatchExplorer-data/pull/23 

Thanks.

matthew-rollitt commented:
This PR adds the ability to automatically test the job/pool templates inside the ncj/ folder. Whenever we have a new rendering image or update the rendering templates these tests will run on every build. If you want to add a new rendering scenario you just a test to the runner/TestConfiguration.json file.

This runner will run on every build and takes roughly 15-25min to run. The runner/README.md file contains more information.
